File: python/wordPattern.py
# Time -> 0(N) Space -> (N)
def wordPattern(pattern: str, s: str) -> bool:
    words = s.split(" ")
    p_map, s_map = {}, {}

    for pat in pattern:
        p_map[pat] = p_map.get(pat, 0) + 1
    for word in words:
        s_map[word] = s_map.get(word, 0) + 1

    w_map, pat_map = {}, {}

    for _, val in p_map.items():
        pat_map[val] = pat_map.get(val, 0) + 1
    for _, val in s_map.items():
        w_map[val] = w_map.get(val, 0) + 1

    if len(w_map) != len(pat_map):
        return False

    for key, val in w_map:
        if pat_map.get(key, 0) == 0 or pat_map[key] != val:
            return False

    return True


# Time -> 0(N) Space -> (N)
def wordPatternII(pattern: str, s: str) -> bool:
    words = s.split(" ")

    if len(pattern) != len(words):
        return False

    letter_map = {}

    for pat in range(len(pattern)):
        curr_pat = pattern[pat]
        if (
            letter_map.get(curr_pat, None) is not None
            and letter_map[curr_pat] != words[pat]
        ):
            return False
        letter_map[curr_pat] = words[pat]

    return True


print(wordPattern("stfj", "ball call car car"))
print(wordPatternII("aba", "dog cat cat"))

File: go/WordPattern/WordPattern.go
package main

import (
	"fmt"
	"strings"
)

// Time -> 0(N) Space -> (N)
func wordPattern(pattern, s string) bool {
	patMap := make(map[byte]int, 0)
	sMap := make(map[string]int, 0)

	words := strings.Split(s, " ")

	for pat := range pattern {
		patMap[pattern[pat]]++
	}

	for word := range words {
		sMap[words[word]]++
	}

	pMap := make(map[int]int, 0)
	wMap := make(map[int]int, 0)

	for _, val := range patMap {
		pMap[val]++
	}

	for _, val := range sMap {
		wMap[val]++
	}

	if len(pMap) != len(wMap) {
		return false
	}

	for key, val := range pMap {
		if wVal, exists := wMap[key]; !exists || wVal != val {
			return false
		}
	}
	return true
}

// Time -> 0(N) Space -> 0(N)
func wordPatternII(pattern, s string) bool {
	letterMap := make(map[byte]string, 0)
	words := strings.Split(s, " ")

	if len(pattern) != len(words) {
		return false
	}

	for pat := range pattern {
		if val, exists := letterMap[pattern[pat]]; exists && val != words[pat] {
			return false
		}
		letterMap[pattern[pat]] = words[pat]
	}
	return true
}

func main() {
	fmt.Println(wordPattern("aba", "dog cat cat"))
	fmt.Println(wordPatternII("aba", "dog cat cat"))
}
